File: api/agent/query_builder.py
"""
Query Builder - Constructs Supabase queries from parsed intents.
"""

from typing import Any, Dict, List, Optional, Tuple

import structlog
from supabase import Client

from api.agent.models import (
    EntityType,
    FilterOperator,
    ParsedIntent,
    QueryFilter,
)

logger = structlog.get_logger()


# Table mappings for each entity type
TABLE_MAPPINGS = {
    EntityType.EMPRESAS: "dim_empresas",
    EntityType.PESSOAS: "dim_pessoas",
    EntityType.NOTICIAS: "dim_noticias",
    EntityType.POLITICOS: "dim_politicos",  # Tabela no brasil-data-hub
}

# Entity types that use external database (fiscal)
EXTERNAL_ENTITIES = {EntityType.POLITICOS}

# Default columns to select for each entity type
SELECT_COLUMNS = {
    EntityType.EMPRESAS: (
        "id, cnpj, razao_social, nome_fantasia, cidade, estado, "
        "cnae_principal, descricao_cnae, porte, regime_tributario, "
        "capital_social, data_abertura, situacao_cadastral, qtd_funcionarios"
    ),
    EntityType.PESSOAS: (
        "id, nome_completo, email, cargo, empresa_id, linkedin_url, telefone"
    ),
    EntityType.NOTICIAS: (
        "id, titulo, resumo, fonte, data_publicacao, segmento, url"
    ),
    # dim_politicos (brasil-data-hub) - Schema real
    EntityType.POLITICOS: (
        "id, cpf, nome_completo, nome_urna, data_nascimento, "
        "sexo, grau_instrucao, ocupacao"
    ),
}

# Valid fields for each entity type (for security)
VALID_FIELDS = {
    EntityType.EMPRESAS: {
        "id", "cnpj", "razao_social", "nome_fantasia", "cidade", "estado",
        "cnae_principal", "descricao_cnae", "porte", "regime_tributario",
        "capital_social", "data_abertura", "situacao_cadastral", "qtd_funcionarios",
        "nome"  # alias for razao_social/nome_fantasia
    },
    EntityType.PESSOAS: {
        "id", "nome_completo", "email", "cargo", "empresa_id", "linkedin_url",
        "telefone", "nome"  # alias for nome_completo
    },
    EntityType.NOTICIAS: {
        "id", "titulo", "resumo", "fonte", "data_publicacao", "segmento", "url",
        "keywords"  # virtual field for searching titulo + resumo
    },
    # dim_politicos (brasil-data-hub) - Schema real
    EntityType.POLITICOS: {
        "id", "cpf", "nome_completo", "nome_urna", "data_nascimento",
        "sexo", "grau_instrucao", "ocupacao",
        "nome",  # alias para nome_completo
        # Campos de mandatos (fato_politicos_mandatos) - buscados via join
        "cargo", "partido_sigla", "partido_nome", "municipio", "codigo_ibge",
        "ano_eleicao", "eleito", "coligacao", "situacao_turno",
    },
}

# Campos que pertencem a fato_politicos_mandatos (não dim_politicos)
MANDATO_FIELDS = {
    "cargo", "partido_sigla", "partido_nome", "municipio", "codigo_ibge",
    "ano_eleicao", "eleito", "coligacao", "situacao_turno", "turno",
    "numero_candidato", "data_inicio_mandato", "data_fim_mandato",
}

# Field aliases (map common names to actual column names)
FIELD_ALIASES = {
    EntityType.EMPRESAS: {
        "nome": "razao_social",
        "name": "razao_social",
        "fantasia": "nome_fantasia",
        "uf": "estado",
        "state": "estado",
        "cnae": "cnae_principal",
        "funcionarios": "qtd_funcionarios",
        "employees": "qtd_funcionarios",
    },
    EntityType.PESSOAS: {
        "nome": "nome_completo",
        "name": "nome_completo",
        "empresa": "empresa_id",
        "linkedin": "linkedin_url",
        "phone": "telefone",
    },
    EntityType.NOTICIAS: {
        "data": "data_publicacao",
        "date": "data_publicacao",
        "source": "fonte",
        "title": "titulo",
        "summary": "resumo",
    },
    # dim_politicos (brasil-data-hub) - Aliases simplificados
    EntityType.POLITICOS: {
        "nome": "nome_completo",
        "name": "nome_completo",
        "nascimento": "data_nascimento",
        "instrucao": "grau_instrucao",
        "escolaridade": "grau_instrucao",
        "partido": "partido_sigla",
        "cidade": "municipio",
    },
}


class QueryBuilder:
    """Builds and executes Supabase queries from parsed intents."""

    def __init__(
        self,
        supabase_client: Client,
        brasil_data_hub_client: Optional[Client] = None,
    ):
        """
        Initialize the query builder.

        Args:
            supabase_client: Supabase client instance (main database)
            brasil_data_hub_client: Optional Supabase client for brasil-data-hub (politicos)
        """
        self.client = supabase_client
        self.brasil_data_hub_client = brasil_data_hub_client

    async def execute(
        self,
        intent: ParsedIntent,
    ) -> Tuple[List[Dict[str, Any]], int]:
        """
        Execute a query based on the parsed intent.

        Args:
            intent: The parsed intent

        Returns:
            Tuple of (results list, total count)
        """
        table_name = TABLE_MAPPINGS.get(intent.entity_type)
        if not table_name:
            logger.error("unknown_entity_type", entity_type=intent.entity_type)
            return [], 0

        select_columns = SELECT_COLUMNS.get(intent.entity_type, "*")

        # Determine which client to use
        if intent.entity_type in EXTERNAL_ENTITIES:
            if not self.brasil_data_hub_client:
                logger.warning("brasil_data_hub_client_not_configured")
                return [], 0
            client = self.brasil_data_hub_client
        else:
            client = self.client

        try:
            # Start building the query
            query = client.table(table_name).select(select_columns, count="exact")

            # Apply filters
            query = self._apply_filters(query, intent)

            # Apply ordering
            if intent.order_by:
                resolved_field = self._resolve_field(intent.entity_type, intent.order_by)
                if resolved_field:
                    query = query.order(resolved_field, desc=intent.order_desc)

            # Apply limit
            query = query.limit(intent.limit)

            # Execute query
            result = query.execute()

            data = result.data or []
            total_count = result.count or len(data)

            logger.info(
                "query_executed",
                entity_type=intent.entity_type,
                filters_count=len(intent.filters),
                results_count=len(data),
                total_count=total_count,
            )

            return data, total_count

        except Exception as e:
            logger.error("query_execution_error", error=str(e))
            return [], 0

    async def fetch_mandatos(
        self,
        politico_id: str,
    ) -> List[Dict[str, Any]]:
        """
        Fetch mandatos history for a specific politician.

        Args:
            politico_id: The UUID of the politician

        Returns:
            List of mandatos records
        """
        if not self.brasil_data_hub_client:
            logger.warning("brasil_data_hub_client_not_configured")
            return []

        try:
            result = self.brasil_data_hub_client.table("fato_politicos_mandatos").select(
                "id, politico_id, codigo_ibge, municipio, cargo, ano_eleicao, "
                "turno, numero_candidato, partido_sigla, partido_nome, coligacao, "
                "situacao_turno, eleito, data_inicio_mandato, data_fim_mandato"
            ).eq("politico_id", politico_id).order("ano_eleicao", desc=True).execute()

            data = result.data or []

            logger.info(
                "mandatos_fetched",
                politico_id=politico_id,
                mandatos_count=len(data),
            )

            return data

        except Exception as e:
            logger.error("mandatos_fetch_error", error=str(e), politico_id=politico_id)
            return []

    async def search_politicos_with_mandatos(
        self,
        intent: ParsedIntent,
    ) -> Tuple[List[Dict[str, Any]], int]:
        """
        Search politicians and enrich with their current/latest mandato info.

        This handles two scenarios:
        1. If filters include mandato fields (cargo, partido, municipio), search
           fato_politicos_mandatos first and join with dim_politicos
        2. Otherwise, search dim_politicos first and enrich with mandato info

        Args:
            intent: The parsed intent

        Returns:
            Tuple of (results list with mandato info, total count)
        """
        if not self.brasil_data_hub_client:
            logger.warning("brasil_data_hub_client_not_configured")
            return [], 0

        # Separate filters: politico fields vs mandato fields
        politico_filters = []
        mandato_filters = []

        for f in intent.filters:
            resolved_field = self._resolve_field(intent.entity_type, f.field)
            if resolved_field and resolved_field in MANDATO_FIELDS:
                mandato_filters.append((resolved_field, f))
            else:
                politico_filters.append(f)

        # If we have mandato filters, search mandatos first
        if mandato_filters:
            return await self._search_via_mandatos(
                mandato_filters, politico_filters, intent.limit
            )

        # Otherwise, search politicos first and enrich
        data, total_count = await self.execute(intent)

        if not data:
            return data, total_count

        # Enrich each politician with their latest mandato
        enriched_data = []
        for politico in data:
            politico_id = politico.get("id")
            if politico_id:
                mandatos = await self.fetch_mandatos(politico_id)
                if mandatos:
                    # Get the most recent mandato (already ordered by ano_eleicao desc)
                    latest = mandatos[0]
                    politico["partido_sigla"] = latest.get("partido_sigla")
                    politico["cargo_atual"] = latest.get("cargo")
                    politico["municipio"] = latest.get("municipio")
                    politico["eleito"] = latest.get("eleito")
                    politico["mandatos_count"] = len(mandatos)
            enriched_data.append(politico)

        return enriched_data, total_count

    async def _search_via_mandatos(
        self,
        mandato_filters: List[Tuple[str, QueryFilter]],
        politico_filters: List[QueryFilter],
        limit: int,
    ) -> Tuple[List[Dict[str, Any]], int]:
        """
        Search politicians by first querying fato_politicos_mandatos.

        Args:
            mandato_filters: Filters for mandato fields (already resolved)
            politico_filters: Filters for politico fields
            limit: Maximum results to return

        Returns:
            Tuple of (results list with mandato info, total count)
        """
        try:
            # Build query on fato_politicos_mandatos with join to dim_politicos
            query = self.brasil_data_hub_client.table("fato_politicos_mandatos").select(
                "id, cargo, partido_sigla, partido_nome, municipio, codigo_ibge, "
                "ano_eleicao, turno, numero_candidato, eleito, coligacao, situacao_turno, "
                "politico:politico_id (id, cpf, nome_completo, nome_urna, "
                "data_nascimento, sexo, grau_instrucao, ocupacao)",
                count="exact",
            )

            # Apply mandato filters
            for resolved_field, f in mandato_filters:
                query = self._apply_single_filter(query, resolved_field, f)

            # Apply politico filters via nested filter
            for f in politico_filters:
                resolved_field = self._resolve_field(EntityType.POLITICOS, f.field)
                if resolved_field and resolved_field not in MANDATO_FIELDS:
                    # For nested politico fields, use the nested filter syntax
                    nested_field = f"politico.{resolved_field}"
                    if f.operator in (FilterOperator.LIKE, FilterOperator.ILIKE):
                        query = query.ilike(nested_field, f"%{f.value}%")
                    elif f.operator == FilterOperator.EQ:
                        query = query.eq(nested_field, f.value)

            # Order by most recent election
            query = query.order("ano_eleicao", desc=True).limit(limit)

            result = query.execute()
            data = result.data or []
            total_count = result.count or len(data)

            # Flatten the data structure
            flattened = []
            seen_politicos = set()

            for mandato in data:
                politico_data = mandato.pop("politico", {}) or {}
                politico_id = politico_data.get("id")

                # Avoid duplicates (same politico with multiple mandatos)
                if politico_id and politico_id in seen_politicos:
                    continue
                seen_politicos.add(politico_id)

                flattened.append({
                    **politico_data,
                    "partido_sigla": mandato.get("partido_sigla"),
                    "cargo_atual": mandato.get("cargo"),
                    "municipio": mandato.get("municipio"),
                    "codigo_ibge": mandato.get("codigo_ibge"),
                    "ano_eleicao": mandato.get("ano_eleicao"),
                    "eleito": mandato.get("eleito"),
                    "coligacao": mandato.get("coligacao"),
                    "situacao_turno": mandato.get("situacao_turno"),
                })

            logger.info(
                "search_via_mandatos_executed",
                mandato_filters_count=len(mandato_filters),
                politico_filters_count=len(politico_filters),
                results_count=len(flattened),
                total_count=total_count,
            )

            return flattened, len(flattened)

        except Exception as e:
            logger.error("search_via_mandatos_error", error=str(e))
            return [], 0

    def _apply_filters(self, query, intent: ParsedIntent):
        """
        Apply filters to the query.

        Args:
            query: The Supabase query builder
            intent: The parsed intent with filters

        Returns:
            The query with filters applied
        """
        for f in intent.filters:
            resolved_field = self._resolve_field(intent.entity_type, f.field)
            if not resolved_field:
                logger.warning("invalid_filter_field", field=f.field)
                continue

            query = self._apply_single_filter(query, resolved_field, f)

        return query

    def _resolve_field(self, entity_type: EntityType, field: str) -> Optional[str]:
        """
        Resolve a field name, handling aliases.

        Args:
            entity_type: The entity type
            field: The field name (possibly an alias)

        Returns:
            The resolved field name or None if invalid
        """
        field_lower = field.lower()

        # Check aliases first
        aliases = FIELD_ALIASES.get(entity_type, {})
        if field_lower in aliases:
            field_lower = aliases[field_lower]

        # Validate field
        valid_fields = VALID_FIELDS.get(entity_type, set())
        if field_lower not in valid_fields:
            return None

        return field_lower

    def _apply_single_filter(self, query, field: str, filter_obj: QueryFilter):
        """
        Apply a single filter to the query.

        Args:
            query: The Supabase query builder
            field: The resolved field name
            filter_obj: The filter object

        Returns:
            The query with the filter applied
        """
        value = filter_obj.value
        operator = filter_obj.operator

        # Handle special virtual fields
        if field == "keywords" and operator in (FilterOperator.LIKE, FilterOperator.ILIKE):
            # Search in titulo and resumo for noticias
            query = query.or_(f"titulo.ilike.%{value}%,resumo.ilike.%{value}%")
            return query

        # Standard operators
        if operator == FilterOperator.EQ:
            query = query.eq(field, value)
        elif operator == FilterOperator.NEQ:
            query = query.neq(field, value)
        elif operator == FilterOperator.GT:
            query = query.gt(field, value)
        elif operator == FilterOperator.GTE:
            query = query.gte(field, value)
        elif operator == FilterOperator.LT:
            query = query.lt(field, value)
        elif operator == FilterOperator.LTE:
            query = query.lte(field, value)
        elif operator in (FilterOperator.LIKE, FilterOperator.ILIKE):
            query = query.ilike(field, f"%{value}%")
        elif operator == FilterOperator.IN:
            if isinstance(value, list):
                query = query.in_(field, value)
        elif operator == FilterOperator.IS_NULL:
            query = query.is_(field, "null")
        elif operator == FilterOperator.NOT_NULL:
            query = query.not_.is_(field, "null")

        return query


# Global query builder (requires supabase client)
def create_query_builder(
    supabase_client: Client,
    brasil_data_hub_client: Optional[Client] = None,
) -> QueryBuilder:
    """
    Create a query builder instance.

    Args:
        supabase_client: Supabase client instance (main database)
        brasil_data_hub_client: Optional Supabase client for brasil-data-hub (politicos)

    Returns:
        QueryBuilder instance
    """
    return QueryBuilder(supabase_client, brasil_data_hub_client)

File: backend/src/atlas/query-builder.js
/**
 * Atlas Query Builder
 * Builds Supabase queries based on parsed intent for political data.
 */

import { createClient } from '@supabase/supabase-js';
import logger from '../utils/logger.js';
import { INTENTS } from './intent-parser.js';

// Brasil Data Hub client (dim_politicos, fato_politicos_mandatos)
const brasilDataHub = process.env.BRASIL_DATA_HUB_URL && process.env.BRASIL_DATA_HUB_KEY
  ? createClient(process.env.BRASIL_DATA_HUB_URL, process.env.BRASIL_DATA_HUB_KEY)
  : null;

/**
 * Check if Brasil Data Hub is configured
 * @returns {boolean}
 */
export function isConfigured() {
  return !!brasilDataHub;
}

/**
 * Search politicians by name
 * @param {Object} entities - Parsed entities
 * @returns {Promise<Object>} - Query results
 */
async function searchPolitician(entities) {
  const { nome } = entities;

  if (!nome) {
    return { error: 'Nome do político não especificado', data: [] };
  }

  // Search in dim_politicos
  const { data: politicos, error } = await brasilDataHub
    .from('dim_politicos')
    .select('id, nome_completo, nome_urna, sexo, ocupacao, grau_instrucao, data_nascimento')
    .or(`nome_completo.ilike.%${nome}%,nome_urna.ilike.%${nome}%`)
    .limit(10);

  if (error) {
    logger.error('Error searching politicians', { error: error.message });
    return { error: error.message, data: [] };
  }

  // Enrich with latest mandato
  const enrichedPoliticians = [];
  for (const politico of politicos || []) {
    const { data: mandatos } = await brasilDataHub
      .from('fato_politicos_mandatos')
      .select('cargo, partido_sigla, partido_nome, municipio, codigo_ibge, ano_eleicao, eleito')
      .eq('politico_id', politico.id)
      .order('ano_eleicao', { ascending: false })
      .limit(3);

    enrichedPoliticians.push({
      ...politico,
      mandatos: mandatos || [],
      ultimo_mandato: mandatos?.[0] || null
    });
  }

  return {
    data: enrichedPoliticians,
    count: enrichedPoliticians.length,
    queryType: 'search_politician'
  };
}

/**
 * Get politician details by ID or name
 * @param {Object} entities - Parsed entities
 * @returns {Promise<Object>} - Query results
 */
async function getPoliticianDetails(entities) {
  const { id, nome } = entities;

  let politico = null;

  if (id) {
    const { data, error } = await brasilDataHub
      .from('dim_politicos')
      .select('id, nome_completo, nome_urna, data_nascimento, sexo, grau_instrucao, ocupacao')
      .eq('id', id)
      .single();

    if (error) {
      logger.error('Error getting politician by ID', { error: error.message, id });
      return { error: 'Político não encontrado', data: null };
    }
    politico = data;
  } else if (nome) {
    // Search and get first match
    const { data, error } = await brasilDataHub
      .from('dim_politicos')
      .select('id, nome_completo, nome_urna, data_nascimento, sexo, grau_instrucao, ocupacao')
      .or(`nome_completo.ilike.%${nome}%,nome_urna.ilike.%${nome}%`)
      .limit(1)
      .single();

    if (error || !data) {
      logger.error('Error getting politician by name', { error: error?.message, nome });
      return { error: 'Político não encontrado', data: null };
    }
    politico = data;
  } else {
    return { error: 'ID ou nome do político não especificado', data: null };
  }

  // Get all mandates
  const { data: mandatos } = await brasilDataHub
    .from('fato_politicos_mandatos')
    .select(`
      id, cargo, partido_sigla, partido_nome, municipio, codigo_ibge,
      ano_eleicao, turno, numero_candidato, eleito, coligacao, situacao_turno,
      data_inicio_mandato, data_fim_mandato
    `)
    .eq('politico_id', politico.id)
    .order('ano_eleicao', { ascending: false });

  return {
    data: {
      politico,
      mandatos: mandatos || [],
      total_mandatos: mandatos?.length || 0,
      partidos_unicos: [...new Set(mandatos?.map(m => m.partido_sigla).filter(Boolean))]
    },
    queryType: 'politician_details'
  };
}

/**
 * Get politicians by party
 * @param {Object} entities - Parsed entities
 * @returns {Promise<Object>} - Query results
 */
async function getPoliticiansByParty(entities) {
  const { partido, cargo, ano_eleicao, eleito } = entities;

  if (!partido) {
    return { error: 'Partido não especificado', data: [] };
  }

  let query = brasilDataHub
    .from('fato_politicos_mandatos')
    .select(`
      id, cargo, partido_sigla, partido_nome, municipio, codigo_ibge,
      ano_eleicao, eleito, situacao_turno,
      politico:politico_id (id, nome_completo, nome_urna, sexo, ocupacao)
    `)
    .eq('partido_sigla', partido.toUpperCase())
    .limit(50);

  if (cargo) {
    query = query.ilike('cargo', `%${cargo}%`);
  }
  if (ano_eleicao) {
    query = query.eq('ano_eleicao', ano_eleicao);
  }
  if (eleito !== undefined && eleito !== null) {
    query = query.eq('eleito', eleito);
  }

  const { data, error, count } = await query.order('ano_eleicao', { ascending: false });

  if (error) {
    logger.error('Error getting politicians by party', { error: error.message, partido });
    return { error: error.message, data: [] };
  }

  // Flatten and deduplicate
  const seen = new Set();
  const politicians = [];
  for (const mandato of data || []) {
    const politico = mandato.politico || {};
    const politicoId = politico.id;
    if (politicoId && !seen.has(politicoId)) {
      seen.add(politicoId);
      politicians.push({
        id: politicoId,
        nome_completo: politico.nome_completo,
        nome_urna: politico.nome_urna,
        sexo: politico.sexo,
        cargo: mandato.cargo,
        municipio: mandato.municipio,
        ano_eleicao: mandato.ano_eleicao,
        eleito: mandato.eleito
      });
    }
  }

  return {
    data: politicians,
    count: politicians.length,
    partido,
    queryType: 'by_party'
  };
}

/**
 * Get politicians by municipality
 * @param {Object} entities - Parsed entities
 * @returns {Promise<Object>} - Query results
 */
async function getPoliticiansByMunicipality(entities) {
  const { municipio, codigo_ibge, cargo, ano_eleicao, eleito } = entities;

  if (!municipio && !codigo_ibge) {
    return { error: 'Município não especificado', data: [] };
  }

  let query = brasilDataHub
    .from('fato_politicos_mandatos')
    .select(`
      id, cargo, partido_sigla, municipio, codigo_ibge,
      ano_eleicao, eleito, situacao_turno,
      politico:politico_id (id, nome_completo, nome_urna, sexo)
    `)
    .limit(100);

  if (codigo_ibge) {
    query = query.eq('codigo_ibge', codigo_ibge);
  } else if (municipio) {
    query = query.ilike('municipio', `%${municipio}%`);
  }

  if (cargo) {
    query = query.ilike('cargo', `%${cargo}%`);
  }
  if (ano_eleicao) {
    query = query.eq('ano_eleicao', ano_eleicao);
  }
  if (eleito !== undefined && eleito !== null) {
    query = query.eq('eleito', eleito);
  }

  const { data, error } = await query.order('cargo');

  if (error) {
    logger.error('Error getting politicians by municipality', { error: error.message, municipio });
    return { error: error.message, data: [] };
  }

  // Flatten data
  const politicians = (data || []).map(mandato => ({
    id: mandato.politico?.id,
    nome_completo: mandato.politico?.nome_completo,
    nome_urna: mandato.politico?.nome_urna,
    sexo: mandato.politico?.sexo,
    cargo: mandato.cargo,
    partido_sigla: mandato.partido_sigla,
    municipio: mandato.municipio,
    codigo_ibge: mandato.codigo_ibge,
    ano_eleicao: mandato.ano_eleicao,
    eleito: mandato.eleito
  }));

  return {
    data: politicians,
    count: politicians.length,
    municipio: municipio || codigo_ibge,
    queryType: 'by_municipality'
  };
}

/**
 * Get statistics about politicians/parties
 * @param {Object} entities - Parsed entities
 * @returns {Promise<Object>} - Query results
 */
async function getStatistics(entities) {
  const { partido, cargo, municipio, ano_eleicao } = entities;

  // Build base query
  let query = brasilDataHub
    .from('fato_politicos_mandatos')
    .select('partido_sigla, cargo, eleito, ano_eleicao, codigo_ibge', { count: 'exact' });

  if (partido) {
    query = query.eq('partido_sigla', partido.toUpperCase());
  }
  if (cargo) {
    query = query.ilike('cargo', `%${cargo}%`);
  }
  if (municipio) {
    query = query.ilike('municipio', `%${municipio}%`);
  }
  if (ano_eleicao) {
    query = query.eq('ano_eleicao', ano_eleicao);
  }

  const { data, error, count } = await query;

  if (error) {
    logger.error('Error getting statistics', { error: error.message });
    return { error: error.message, data: {} };
  }

  // Calculate statistics
  const stats = {
    total_mandatos: count,
    total_eleitos: data?.filter(m => m.eleito === true).length || 0,
    por_partido: {},
    por_cargo: {},
    por_ano: {}
  };

  for (const mandato of data || []) {
    // By party
    const p = mandato.partido_sigla || 'SEM PARTIDO';
    stats.por_partido[p] = (stats.por_partido[p] || 0) + 1;

    // By position
    const c = mandato.cargo || 'OUTRO';
    stats.por_cargo[c] = (stats.por_cargo[c] || 0) + 1;

    // By year
    const y = mandato.ano_eleicao;
    if (y) {
      stats.por_ano[y] = (stats.por_ano[y] || 0) + 1;
    }
  }

  // Sort parties by count
  stats.partidos_ordenados = Object.entries(stats.por_partido)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 10);

  return {
    data: stats,
    queryType: 'statistics'
  };
}

/**
 * List all parties with counts
 * @returns {Promise<Object>} - Query results
 */
async function listParties() {
  const { data, error } = await brasilDataHub
    .from('fato_politicos_mandatos')
    .select('partido_sigla, partido_nome')
    .not('partido_sigla', 'is', null);

  if (error) {
    logger.error('Error listing parties', { error: error.message });
    return { error: error.message, data: [] };
  }

  // Count by party
  const partyCounts = {};
  const partyNames = {};
  for (const row of data || []) {
    const sigla = row.partido_sigla;
    partyCounts[sigla] = (partyCounts[sigla] || 0) + 1;
    if (row.partido_nome) {
      partyNames[sigla] = row.partido_nome;
    }
  }

  // Convert to array and sort
  const parties = Object.entries(partyCounts)
    .map(([sigla, count]) => ({
      sigla,
      nome: partyNames[sigla] || sigla,
      count
    }))
    .sort((a, b) => b.count - a.count);

  return {
    data: parties,
    count: parties.length,
    queryType: 'party_list'
  };
}

/**
 * Execute query based on intent
 * @param {Object} parsedIntent - Result from intent parser
 * @returns {Promise<Object>} - Query results
 */
export async function executeQuery(parsedIntent) {
  const startTime = Date.now();
  const { intent, entities } = parsedIntent;

  if (!brasilDataHub) {
    return {
      error: 'Brasil Data Hub não configurado. Configure BRASIL_DATA_HUB_URL e BRASIL_DATA_HUB_KEY.',
      data: null
    };
  }

  let result;

  switch (intent) {
    case INTENTS.SEARCH_POLITICIAN:
      result = await searchPolitician(entities);
      break;

    case INTENTS.POLITICIAN_DETAILS:
      result = await getPoliticianDetails(entities);
      break;

    case INTENTS.BY_PARTY:
      result = await getPoliticiansByParty(entities);
      break;

    case INTENTS.BY_MUNICIPALITY:
      result = await getPoliticiansByMunicipality(entities);
      break;

    case INTENTS.STATISTICS:
      result = await getStatistics(entities);
      break;

    case INTENTS.PARTY_LIST:
      result = await listParties();
      break;

    case INTENTS.FOLLOW_UP:
      // Re-execute with merged entities from context
      // The orchestrator should have merged the entities already
      if (entities.id || entities.nome) {
        result = await getPoliticianDetails(entities);
      } else if (entities.partido) {
        result = await getPoliticiansByParty(entities);
      } else if (entities.municipio) {
        result = await getPoliticiansByMunicipality(entities);
      } else {
        result = await getStatistics(entities);
      }
      break;

    case INTENTS.GENERAL:
    default:
      // Try to infer from entities
      if (entities.nome) {
        result = await searchPolitician(entities);
      } else if (entities.partido) {
        result = await getPoliticiansByParty(entities);
      } else if (entities.municipio) {
        result = await getPoliticiansByMunicipality(entities);
      } else {
        result = await listParties();
      }
      break;
  }

  const processingTime = Date.now() - startTime;
  logger.debug('Query executed', {
    intent,
    queryType: result.queryType,
    resultCount: Array.isArray(result.data) ? result.data.length : 1,
    processingTime
  });

  return {
    ...result,
    processingTime
  };
}

export default {
  executeQuery,
  isConfigured
};
